test/utils/controller/controller.go

```diff
@@ -0,0 +1,48 @@
+/*
```

Comment:
Apologies for not getting the point, but is it kind of defeat the purpose of integration tests if we use this mock? Is there any reason why we cannot use the Controller with the API server for testing?

Reply:
The clusterschedulingpolicysnapshot integration test is to test the real snapshot controller only. The util is to fake its dependency, crp controller.

we can add integration tests for the whole crp and its related watcher(dependency) in test/integration location. And we could use the real controller implementation.

Reply:
Oh, I mean, in this case, `NewController` with a no-op keyFunc/reconcileFunc would also work, right?

Reply:
For this one I don't have a strong preference though; a little bit more real would be better, but I guess it is true that the yield diminishes very quickly.